File: clients/android/app/src/main/java/ai/vocify/vellumassistant/WorkOSAuth.java
package ai.vocify.vellumassistant;

import android.net.Uri;
import android.util.Base64;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

final class WorkOSAuth {
    static final String API_BASE_URL = "https://api.workos.com";

    private static final String PROVIDER_ID = "workos";
    private static final String SCOPE = "openid profile email";
    private static final SecureRandom SECURE_RANDOM = new SecureRandom();

    private WorkOSAuth() {}

    static String generateBase64UrlToken() {
        byte[] bytes = new byte[32];
        SECURE_RANDOM.nextBytes(bytes);
        return base64UrlEncode(bytes);
    }

    static String codeChallenge(String verifier) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            return base64UrlEncode(digest.digest(verifier.getBytes(StandardCharsets.UTF_8)));
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 is unavailable", e);
        }
    }

    static Uri buildAuthorizeUri(
        String clientId,
        String redirectUri,
        String challenge,
        String state,
        String loginHint,
        String providerHint,
        String intent
    ) {
        String provider = nonEmpty(providerHint);
        Uri.Builder builder = Uri.parse(API_BASE_URL)
            .buildUpon()
            .appendPath("user_management")
            .appendPath("authorize")
            .appendQueryParameter("client_id", clientId)
            .appendQueryParameter("redirect_uri", redirectUri)
            .appendQueryParameter("response_type", "code")
            .appendQueryParameter("scope", SCOPE)
            .appendQueryParameter("code_challenge", challenge)
            .appendQueryParameter("code_challenge_method", "S256")
            .appendQueryParameter("state", state)
            .appendQueryParameter("provider", provider != null ? provider : "authkit");

        String email = nonEmpty(loginHint);
        if (email != null) {
            builder.appendQueryParameter("login_hint", email);
        }
        if ("signup".equals(intent)) {
            builder.appendQueryParameter("screen_hint", "sign-up");
        }
        return builder.build();
    }

    static String selectClientId(String configBody) throws JSONException {
        JSONObject social = new JSONObject(configBody)
            .getJSONObject("data")
            .getJSONObject("socialaccount");
        JSONArray providers = social.getJSONArray("providers");

        for (int index = 0; index < providers.length(); index++) {
            JSONObject provider = providers.getJSONObject(index);
            if (!provider.isNull("openid_configuration_url")) {
                continue;
            }
            if (!hasFlow(provider.optJSONArray("flows"), "provider_token")) {
                continue;
            }
            String clientId = nonEmpty(provider.optString("client_id", null));
            if (clientId != null) {
                return clientId;
            }
        }
        return null;
    }

    static JSONObject authenticateRequestBody(String clientId, String code, String verifier)
        throws JSONException {
        JSONObject body = new JSONObject();
        body.put("client_id", clientId);
        body.put("grant_type", "authorization_code");
        body.put("code", code);
        body.put("code_verifier", verifier);
        return body;
    }

    static String accessToken(String authenticateBody) throws JSONException {
        return nonEmpty(new JSONObject(authenticateBody).optString("access_token", null));
    }

    static JSONObject providerTokenRequestBody(String clientId, String accessToken)
        throws JSONException {
        JSONObject token = new JSONObject();
        token.put("client_id", clientId);
        token.put("access_token", accessToken);

        JSONObject body = new JSONObject();
        body.put("provider", PROVIDER_ID);
        body.put("process", "login");
        body.put("token", token);
        return body;
    }

    static String sessionToken(String providerTokenBody) throws JSONException {
        return nonEmpty(new JSONObject(providerTokenBody)
            .getJSONObject("meta")
            .optString("session_token", null));
    }

    private static String base64UrlEncode(byte[] bytes) {
        return Base64.encodeToString(bytes, Base64.URL_SAFE | Base64.NO_WRAP | Base64.NO_PADDING);
    }

    private static boolean hasFlow(JSONArray flows, String needle) throws JSONException {
        if (flows == null) {
            return false;
        }
        for (int index = 0; index < flows.length(); index++) {
            if (needle.equals(flows.optString(index))) {
                return true;
            }
        }
        return false;
    }

    private static String nonEmpty(String value) {
        return value == null || value.isEmpty() ? null : value;
    }
}

File: clients/ios/App/App/WorkOSAuth.swift
import CryptoKit
import Foundation

/// Pure helpers for PKCE flow against WorkOS.
enum WorkOSAuth {
    static let apiBaseURL = "https://api.workos.com"
    static let providerId = "workos"
    static let scope = "openid profile email"

    /// base64url without padding (RFC 7636 §A).
    static func base64URLEncode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    /// 32 random bytes, base64url-encoded. Returns nil on RNG failure,
    /// which callers must treat as fatal.
    static func generateCodeVerifier() -> String? {
        var bytes = [UInt8](repeating: 0, count: 32)
        guard SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess
        else { return nil }
        return base64URLEncode(Data(bytes))
    }

    static func codeChallenge(forVerifier verifier: String) -> String {
        let digest = SHA256.hash(data: Data(verifier.utf8))
        return base64URLEncode(Data(digest))
    }

    /// Build the WorkOS authorize URL.
    /// `prompt` is deliberately omitted for parity with other clients,
    /// but we current spawn an ephemeral session anyway.
    static func buildAuthorizeURL(
        clientId: String,
        redirectURI: String,
        challenge: String,
        state: String,
        loginHint: String?,
        intent: String?
    ) -> URL? {
        var components = URLComponents(string: "\(apiBaseURL)/user_management/authorize")
        var items = [
            URLQueryItem(name: "client_id", value: clientId),
            URLQueryItem(name: "redirect_uri", value: redirectURI),
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "scope", value: scope),
            URLQueryItem(name: "code_challenge", value: challenge),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "provider", value: "authkit"),
        ]
        if let loginHint = nonEmpty(loginHint) {
            items.append(URLQueryItem(name: "login_hint", value: loginHint))
        }
        if intent == "signup" {
            items.append(URLQueryItem(name: "screen_hint", value: "sign-up"))
        }
        components?.queryItems = items
        return components?.url
    }

    /// Pick the OAuth2 WorkOS provider's client id.
    /// During the transition window, two entries share the `workos-oidc` id.
    /// The usable one has token auth and no OIDC discovery URL.
    /// Returns nil if there is no compatible token provider, in which
    /// callers should surface a clear error.
    static func selectClientId(fromConfig data: Data) -> String? {
        guard
            let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let dataObj = root["data"] as? [String: Any],
            let social = dataObj["socialaccount"] as? [String: Any],
            let providers = social["providers"] as? [[String: Any]]
        else { return nil }

        for provider in providers {
            // `as? String == nil` treats an absent key, JSON null (NSNull
            // under JSONSerialization), and any non-string alike as "no
            // discovery URL" — i.e. the OAuth2 provider, not legacy OIDC.
            guard (provider["openid_configuration_url"] as? String) == nil,
                  let flows = provider["flows"] as? [String],
                  flows.contains("provider_token"),
                  let clientId = provider["client_id"] as? String
            else { continue }
            return clientId
        }
        return nil
    }

    static func authenticateRequestBody(clientId: String, code: String, verifier: String) -> Data? {
        try? JSONSerialization.data(withJSONObject: [
            "client_id": clientId,
            "grant_type": "authorization_code",
            "code": code,
            "code_verifier": verifier,
        ])
    }

    static func accessToken(fromAuthenticate data: Data) -> String? {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return nonEmpty(json["access_token"] as? String)
    }

    static func providerTokenRequestBody(clientId: String, accessToken: String) -> Data? {
        try? JSONSerialization.data(withJSONObject: [
            "provider": providerId,
            "process": "login",
            "token": ["client_id": clientId, "access_token": accessToken],
        ])
    }

    static func sessionToken(fromProviderToken data: Data) -> String? {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let meta = json["meta"] as? [String: Any]
        else { return nil }
        return nonEmpty(meta["session_token"] as? String)
    }

    private static func nonEmpty(_ value: String?) -> String? {
        guard let value = value, !value.isEmpty else { return nil }
        return value
    }
}
